grails-app/controllers/org/pih/warehouse/api/PrepaymentInvoiceItemApiController.groovy

```diff
@@ -23,8 +23,13 @@ class PrepaymentInvoiceItemApiController {
 
     def update() {
         JSONObject jsonObject = request.JSON
-        Integer quantity = jsonObject.get('quantity')
-        prepaymentInvoiceService.updateInvoiceItemQuantity(params.id, quantity)
+        Integer quantity = jsonObject.get('quantity', null)
+        if (quantity != null) {
+            prepaymentInvoiceService.updateInvoiceItemQuantity(params.id, quantity)
+        } else {
+            BigDecimal unitPrice = jsonObject.get('unitPrice')
+            prepaymentInvoiceService.updateInvoiceItemUnitPrice(params.id, unitPrice)
+        }
```

Comment:
maybe it will look cleaner if you create a method like "updateInvoiceItem"
and within this method decide whether you want to update item quantity or unit price, instead of making this decision in the controller. In case we want to make more updates, we will have to nest thousands of if-else closures here.

Reply:
will check that

Reply:
I'd either make two methods: `updateInvoiceItemForOrderItem` and `updateInvoiceItemForAdjustment` that each have their separate flows (but both stay general purpose so that our controllers can stay clean like Alan said), or just make a single `updateInvoiceItem` like Alan said.

I know you're doing a lot of validation in the service for both update scenarios and so merging the two could be messy, but maybe that's a sign that we're better off with a command object that has its own validation step that we can define. Something like:
```
InvoiceItemCommand command = new InvoiceItemCommand(params)
command.validate()
if (command.hasErrors())
    // error early, before we even try to go to the db

InvoiceItem invoiceItemToUpdate = InvoiceItem.findById(command.id)
invoiceItemToUpdate.unitPrice = command.unitPrice
invoiceItemToUpdate.save()
if (invoiceItemToUpdate.hasErrors())
    // error
else
    // return normally
```
Or maybe we need a custom InvoiceItemValidator class that encapsulates the more complex validation.


Sorry if this is overkill. I've been reading up about request validation the past couple days. Seems like a topic we could think about how we want to handle in general https://docs.grails.org/latest/guide/validation.html#:~:text=A%20common%20pattern%20in%20Grails,share%20properties%20and%20their%20constraints.

Reply:
Yeah actually a thought came to my mind about the command with validate today. That could be useful, but I wonder if we could do it in a cleanup/improvement ticket or something like that? So we get this feature to the QA asap, and then we can focus on all those improvements we wanted to have here like: 
1. command with validation
2. improvements on the transients with sums
3. moving to transients everything that can be moved
4. pulling out some common logic to some service methods
5. more unit tests
6. etc (anything more that comes to your mind right away?)

@EWaterman @alannadolny what do you guys think?

Reply:
yeah that seems reasonable to me. I'll think more about it today.